### ISSUE TITLE
CorsPolicyBuilder does not correctly combine policy with DefaultIsOriginAllowed

### ISSUE THREAD

**Issue by dIeGoLi** [dotnet/aspnetcore#43319]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

A CorsPolicy has per default a DefaultIsOriginAllowed func which is responsible for evaluating if an origin is allowed. Because this func [encapsulates a member access to `Origins`](https://github.com/dotnet/aspnetcore/blob/7b5da426f9bd42fa515a6b67e0bd1504eacbc6c7/src/Middleware/CORS/src/Infrastructure/CorsPolicy.cs#L177), cors policy builder will not correctly combine two cors policies.
Following sample code demonstrates the problem. It is the Program.cs of a new .net 6 Web API Project.

```
using Microsoft.AspNetCore.Cors.Infrastructure;

var builder = WebApplication.CreateBuilder(args);

// Add services to the container.

builder.Services.AddControllers();
// Learn more about configuring Swagger/OpenAPI at https://aka.ms/aspnetcore/swashbuckle
builder.Services.AddEndpointsApiExplorer();
builder.Services.AddSwaggerGen();

// does not work
var policy1 = new CorsPolicyBuilder(new CorsPolicy())
  .AllowAnyHeader()
  .AllowAnyMethod()
  .WithOrigins("https://www.microsoft.com")
  .Build();
//builder.Services.AddCors(options => {
//  options.AddDefaultPolicy(policy1);
//});

// works
var policy2 = new CorsPolicyBuilder("https://www.microsoft.com")
  .AllowAnyHeader()
  .AllowAnyMethod()
  .Build();

// will not work if cors request is made for origin https://www.visualstudio.com
//var policy3 = new CorsPolicyBuilder(policy2)
//  .WithOrigins("https://www.visualstudio.com")
//  .Build();


builder.Services.AddCors(options => {
  options.AddDefaultPolicy(policy2);
});

Console.WriteLine(policy1.IsOriginAllowed("https://www.microsoft.com")); // => false
Console.WriteLine(policy2.IsOriginAllowed("https://www.microsoft.com")); // => true

var app = builder.Build();

// Configure the HTTP request pipeline.
if (app.Environment.IsDevelopment()) {
  app.UseSwagger();
  app.UseSwaggerUI();
}

app.UseHttpsRedirection();
app.UseCors();
app.UseAuthorization();
app.MapControllers();
app.Run();
```

### Expected Behavior

Combining two policies using [`CorsPolicyBuilder`](https://github.com/dotnet/aspnetcore/blob/7b5da426f9bd42fa515a6b67e0bd1504eacbc6c7/src/Middleware/CORS/src/Infrastructure/CorsPolicyBuilder.cs#L29) where each policy has allowed origins should combine correctly. 
One solution would be, that if the first policy has `IsDefaultIsOriginAllowed` set to true then the IsOriginAllowed func of the second policy should be kept in the [Combine](https://github.com/dotnet/aspnetcore/blob/7b5da426f9bd42fa515a6b67e0bd1504eacbc6c7/src/Middleware/CORS/src/Infrastructure/CorsPolicyBuilder.cs#L260) function. A prettier solution would probably be to consider changes around the IsDefaultIsOriginAllowed / IsOriginAllowed  func approach, but might be more difficult to implement without breaking changes.

### .NET Version

6.0.203

**Comment by captainsafia** [comment]
Is the bug you've identified with the lines of code here?

```
Console.WriteLine(policy1.IsOriginAllowed("https://www.microsoft.com")); // => false
Console.WriteLine(policy2.IsOriginAllowed("https://www.microsoft.com")); // => true
```

One thing to note about the code sample above is that `policy2` is the only Cors policy registered in the same code below and that `IsOriginAllowed` is called before any of the conventions on the policies have been applied.

**Comment by BrennanConroy** [comment]
Taking a glance at the CORS code and the code posted above, I think they are saying the problem code is 
```
new CorsPolicyBuilder(new CorsPolicy())
.WithOrigins("https://www.microsoft.com")
```
and if you remove `new CorsPolicy()` then it will work.

**Comment by halter73** [comment]
I see. The following is setting the `policy.IsOriginAllowed` `Func<string, bool>` which is based on the old policy's origins even though IsOriginAllowed was never changed from the default. While powerful, having a settable `Func` for this is what made it so easy to introduce this bug.

I'm guessing this:

https://github.com/dotnet/aspnetcore/blob/00ed4bb71ffd4b5527594ab0703e5d7ed6c26afa/src/Middleware/CORS/src/Infrastructure/CorsPolicyBuilder.cs#L249-L260

Should be this:

```csharp
    private CorsPolicyBuilder Combine(CorsPolicy policy)
    {
        if (policy == null)
        {
            throw new ArgumentNullException(nameof(policy));
        }

        WithOrigins(policy.Origins.ToArray());
        WithHeaders(policy.Headers.ToArray());
        WithExposedHeaders(policy.ExposedHeaders.ToArray());
        WithMethods(policy.Methods.ToArray());

		if (!policy.IsDefaultIsOriginAllowed)
        {
              SetIsOriginAllowed(policy.IsOriginAllowed);
        }
```

**Comment by dIeGoLi** [comment]
@halter73 exactly

As a sidenote. When combining two policies where each (or at least the first) has not the default `IsOriginAllowed` func, could be considered to throw an exception. Because with the current design of CorsPolicyBuilder it is impossible to combine such two policies. It would only work, if instead of combining the two policies, they would be chained, but i don't think that's a practical use case.

**Comment by dIeGoLi** [comment]
@captainsafia Yes basically the two console writelines demonstrate the problem. I added _Steps To Reproduce_ to clarify the issue.
